### PR TITLE
db: support range keys in delete-only compactions

### PR DESCRIPTION
**Pull request by nicktrav**
~NOTE: This PR depends on #1759.~

Two commits in this patch - the first defines a new composite iterator, the second puts it to use.

This is a re-do of #1656, rebased on latest master to pick up various interfaces changes, and also squash all of the commits from that PR. All feedback from the original PR has been addressed.

TODOs:
- [x] Land #1759
- [x] Rebase on master to pick up compaction support for range keys
- [x] Squash extra commits

---

**db: add tableRangedDeletionIter for iteration of ranged deletion spans**

Add a new `keyspan.FragmentIterator` implementation that iterates over
the range deletion and range key blocks of a single table. The iterator
returns merged spans from the table corresponding to "ranged deletions".
A "ranged deletion" span contains keys that fall into one of three
cases:

- the span contains only range deletions,
- the span contains only range key deletions, or
- the span contains a mixture of both range deletions and range key
  deletions.

Under the hood, the iterator first defragments the range deletion and
range key deletion blocks separately, using the table's respective
`keyspan.FragmentIterator`s. This defragmentation process also filters
out non-deletion key types (i.e. range key SET and UNSETS). The two
separate iterators are then combined into a `keyspan.MergingIter`.

This new `tableRangedDeletionIter` will be used to construct spans
necessary for calculating deletion hints on a table that takes range
keys into account. Previously, only range deletion spans were
considered.

For table that do not contain range key deletes, the iterator returns
the same set of spans that the existing logic would have returned.

**db: consider range keys in delete-only compactions**

Currently, when constructing delete-only compactions hints, only range
delete spans are considered. With the introduction of range keys, the
current hint construction logic is incorrect. Specifically, a hint
constructed from a range delete may completely cover a table containing
range keys. Currently, this table would be marked as eligible for
deletion. However, the table should only be deleted if the range keys
are also deleted.

Make use of the `tableRangedDeletionIter` for construction of
delete-only compaction hints. A new `deleteCompactionHintType` type is
used to distinguish between hints generated by a range delete (which may
not delete tables containing range keys), hints generated from a range
key delete (which may not delete tables containing point keys), and
hints generated from a span with both a range delete and range key
delete (which may delete any covering table).

### CODE REVIEW

File: table_stats.go
Comment (nicktrav) [inline table_stats.go:645]: Bummer. Turns out in versions of Go prior to 1.18 these can't be inlined. Will amend.

Comment (nicktrav) [reply table_stats.go]: Actually ... that seems to be fake news. Interestingly, what I had originally compiles just fine for me locally, but it's failing in CI. Lifting the functions out didn't do anything either.

Comment (nicktrav) [reply table_stats.go]: Fixed. Dodgy rebase didn't pick up the `keyspan.Span` to `*keyspan.Span` change.

**Review by jbowens — COMMENTED** [review]
looking good, thanks for doing this

> Reviewed 1 of 6 files at r1, 8 of 8 files at r3, 4 of 5 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1750#-:-N4D81Ve9JJeb2GFqWNN:bjxx082)* status: 8 of 9 files reviewed, 6 unresolved discussions (waiting on @itsbilal, @jbowens, and @nicktrav)

___
*[`table_stats.go` line 313 at r3](https://reviewable.io/reviews/cockroachdb/pebble/1750#-N4D4n5s-aIQrqL6Su53:-N4D4n5s-aIQrqL6Su54:b-3ivpba) ([raw file](https://github.com/cockroachdb/pebble/blob/c2830ade5b395bc3422277c9eb1d4bbc3585f1b3/table_stats.go#L313)):*
> ```Go
> 	defer iter.Close()
> 	var compactionHints []deleteCompactionHint
> 	// We iterate over the defragmented range tombstones, which ensures
> ```

nit: s/defragmented range tombstones/defragmented range tombstones and range key deletions/
___
*[`table_stats.go` line 338 at r3](https://reviewable.io/reviews/cockroachdb/pebble/1750#-N4D53SgElfz3lTWjfNk:-N4D53SgElfz3lTWjfNl:bydw9co) ([raw file](https://github.com/cockroachdb/pebble/blob/c2830ade5b395bc3422277c9eb1d4bbc3585f1b3/table_stats.go#L338)):*
> ```Go
> 			return nil, err
> 		}
> 		stats.RangeDeletionsBytesEstimate += estimate
> ```

here, and above in the `level == numLevels - 1` conditional, I think we only want to populate `RangeDeletionsBytesEstimate` if `s` contains any keys of kind `RANGEDEL`.

`RangeDeletionsBytesEstimate` is the estimate of bytes within `[start, end)` in _point key_ blocks, or if tables are wholly contained, the entirety of the sstable. We want to avoid totaling this if there's only a `RANGEKEYDEL` over the span.

We should also leave a comment that if `s` only contains `RANGEDEL` and the span wholly contains a table containing range keys, this estimate will be off, mistakenly including the size of the range key block. That's expected to be okay, because range keys are expected to be rare and not contribute significantly to sstable size.
___
*[`table_stats.go` line 584 at r3](https://reviewable.io/reviews/cockroachdb/pebble/1750#-N4D2aSE9Jii5PemLWY4:-N4D2aSE9Jii5PemLWY5:b-k246zt) ([raw file](https://github.com/cockroachdb/pebble/blob/c2830ade5b395bc3422277c9eb1d4bbc3585f1b3/table_stats.go#L584)):*
> ```Go
> // corresponding to the largest and smallest sequence numbers encountered across
> // the range deletes and range keys deletes that comprised the merged spans.
> type tableRangedDeletionIter struct {
> ```

nit: do we need this type now? or could we return the merging iter as an opaque `FragmentIterator`?
___
*[`table_stats.go` line 589 at r3](https://reviewable.io/reviews/cockroachdb/pebble/1750#-N4D40Q_5rlEOaR_rIuI:-N4D40Q_5rlEOaR_rIuJ:bx6p8j4) ([raw file](https://github.com/cockroachdb/pebble/blob/c2830ade5b395bc3422277c9eb1d4bbc3585f1b3/table_stats.go#L589)):*
> ```Go
> 
> // newTableRangeDeletionIter returns a new tableRangedDeletionIter.
> func newTableRangeDeletionIter(
> ```

nit: can we call this something that makes it a little clear that it's range deletions and range key deletions? `newTableRangedDeletionIter` is better, maybe `newCombinedDeletionKeyspanIter` or something is even better? The term 'range deletion' is just so ubiquitous referring specifically to `RANGEDEL` that anything close to it is likely to be confused.
___
*[`table_stats.go` line 674 at r3](https://reviewable.io/reviews/cockroachdb/pebble/1750#-N4D1w2FD7aSCjaxXK1k:-N4D1w2FD7aSCjaxXK1l:b-5qwpvj) ([raw file](https://github.com/cockroachdb/pebble/blob/c2830ade5b395bc3422277c9eb1d4bbc3585f1b3/table_stats.go#L674)):*
> ```Go
> 	// The separate iters for the range dels and range keys are wrapped in a
> 	// merging iter to join the keyspaces into a single keyspace.
> 	mIter := &keyspan.MergingIter{}
> ```

nit:  now that we have [`MergingIter.AddLevel`](https://github.com/cockroachdb/pebble/blob/76a1fffea1e69d969c5c9e6a83ce96e92a8aab96/internal/keyspan/merging_iter.go#L325), we can avoid the allocations of appending to the `[]FragmentIterator` slice by `Init`-ing the `MergingIter` higher up, and directly calling `AddLevel` when we find a non-nil rangedel or range key iterator.

**Review by nicktrav — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1750)* status: 5 of 9 files reviewed, 1 unresolved discussion (waiting on @itsbilal and @jbowens)

___
*[`table_stats.go` line 313 at r3](https://reviewable.io/reviews/cockroachdb/pebble/1750#-N4D4n5s-aIQrqL6Su53:-N4D_sFx9Uz1Znsl3VR3:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/c2830ade5b395bc3422277c9eb1d4bbc3585f1b3/table_stats.go#L313)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

nit: s/defragmented range tombstones/defragmented range tombstones and range key deletions/
</blockquote></details>

Done.
___
*[`table_stats.go` line 338 at r3](https://reviewable.io/reviews/cockroachdb/pebble/1750#-N4D53SgElfz3lTWjfNk:-N4Dh2BD9nJClXmnJfDX:b-239gb6) ([raw file](https://github.com/cockroachdb/pebble/blob/c2830ade5b395bc3422277c9eb1d4bbc3585f1b3/table_stats.go#L338)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

here, and above in the `level == numLevels - 1` conditional, I think we only want to populate `RangeDeletionsBytesEstimate` if `s` contains any keys of kind `RANGEDEL`.

`RangeDeletionsBytesEstimate` is the estimate of bytes within `[start, end)` in _point key_ blocks, or if tables are wholly contained, the entirety of the sstable. We want to avoid totaling this if there's only a `RANGEKEYDEL` over the span.

We should also leave a comment that if `s` only contains `RANGEDEL` and the span wholly contains a table containing range keys, this estimate will be off, mistakenly including the size of the range key block. That's expected to be okay, because range keys are expected to be rare and not contribute significantly to sstable size.
</blockquote></details>

Turns out there was a bug here that wasn't calculating stats for tables with just range keys. Fixed and added an additional test case.
___
*[`table_stats.go` line 584 at r3](https://reviewable.io/reviews/cockroachdb/pebble/1750#-N4D2aSE9Jii5PemLWY4:-N4DaZya42JSEOlYi6Co:b140nx2) ([raw file](https://github.com/cockroachdb/pebble/blob/c2830ade5b395bc3422277c9eb1d4bbc3585f1b3/table_stats.go#L584)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

nit: do we need this type now? or could we return the merging iter as an opaque `FragmentIterator`?
</blockquote></details>

Good call. The `new` function now returns the interface.
___
*[`table_stats.go` line 589 at r3](https://reviewable.io/reviews/cockroachdb/pebble/1750#-N4D40Q_5rlEOaR_rIuI:-N4DbDdC85XSsvNn6tZ9:bg2zg7u) ([raw file](https://github.com/cockroachdb/pebble/blob/c2830ade5b395bc3422277c9eb1d4bbc3585f1b3/table_stats.go#L589)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

nit: can we call this something that makes it a little clear that it's range deletions and range key deletions? `newTableRangedDeletionIter` is better, maybe `newCombinedDeletionKeyspanIter` or something is even better? The term 'range deletion' is just so ubiquitous referring specifically to `RANGEDEL` that anything close to it is likely to be confused.
</blockquote></details>

I think I'd just typo'd originally - intention was for `newTableRangedDeletionIter`. That highlights the problem, so let's go with `newCombinedDeletionKeyspanIter`.
___
*[`table_stats.go` line 674 at r3](https://reviewable.io/reviews/cockroachdb/pebble/1750#-N4D1w2FD7aSCjaxXK1k:-N4F8f5D2vdEcZWW5X7O:bx06w0k) ([raw file](https://github.com/cockroachdb/pebble/blob/c2830ade5b395bc3422277c9eb1d4bbc3585f1b3/table_stats.go#L674)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

nit:  now that we have [`MergingIter.AddLevel`](https://github.com/cockroachdb/pebble/blob/76a1fffea1e69d969c5c9e6a83ce96e92a8aab96/internal/keyspan/merging_iter.go#L325), we can avoid the allocations of appending to the `[]FragmentIterator` slice by `Init`-ing the `MergingIter` higher up, and directly calling `AddLevel` when we find a non-nil rangedel or range key iterator.
</blockquote></details>

Nice! Updated.

**Review by itsbilal — APPROVED** [review]
:lgtm: great work!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1750#-:-N4ZBngE0ie2bBlmxXvo:b-2zonaq)* status: 5 of 9 files reviewed, 4 unresolved discussions (waiting on @jbowens and @nicktrav)

___
*[`table_stats.go` line 313 at r3](https://reviewable.io/reviews/cockroachdb/pebble/1750#-N4Z-xvY-j4InlgPPftK:-N4Z-xvZ-JYUSyHvzXAh:b-jzxobk) ([raw file](https://github.com/cockroachdb/pebble/blob/c2830ade5b395bc3422277c9eb1d4bbc3585f1b3/table_stats.go#L313)):*
> ```Go
> 	defer iter.Close()
> 	var compactionHints []deleteCompactionHint
> 	// We iterate over the defragmented range tombstones, which ensures
> ```

This would be range key deletions too, right? Not just range tombstones?
___
*[`testdata/table_stats` line 264 at r3](https://reviewable.io/reviews/cockroachdb/pebble/1750#-N4ZBCSwC9IFMORNgmht:-N4ZBCSwC9IFMORNgmhu:b93h509) ([raw file](https://github.com/cockroachdb/pebble/blob/c2830ade5b395bc3422277c9eb1d4bbc3585f1b3/testdata/table_stats#L264)):*
> ```
> 
> # Ingest a table with range keys.
> # TODO(travers): Once range keys in batches are flushed to tables, this testcase
> ```

This should be the case now, right?
___
*[`testdata/table_stats_deletion_iter` line 57 at r3](https://reviewable.io/reviews/cockroachdb/pebble/1750#-N4ZBOYy7g6VoQjaJnh7:-N4ZBOYy7g6VoQjaJnh8:bpksncc) ([raw file](https://github.com/cockroachdb/pebble/blob/c2830ade5b395bc3422277c9eb1d4bbc3585f1b3/testdata/table_stats_deletion_iter#L57)):*
> ```
> 
> build
> a-z:{(#0,RANGEKEYDEL) (#0,RANGEKEYSET,@1,foo) (#0,RANGEKEYUNSET,@2)}
> ```

Maybe we should have a case where the RANGEKEYDEL has a lower sequence number than the others to confirm this is the Filter working and not the deletion shadowing the others?

**Review by nicktrav — COMMENTED** [review]
Rebased on master to pick up the recent compaction changes, and pushed a few more commits addressing feedback. I'm also doing a burn of the metamorphic tests.

I'll wait for one more LGTM before squashing, landing, and putting this one to bed ... finally.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1750#-:-N4dHyAb3UPN-5bE5SbG:bgpatcq)* status: 1 of 9 files reviewed, 4 unresolved discussions (waiting on @itsbilal and @jbowens)

___
*[`table_stats.go` line 313 at r3](https://reviewable.io/reviews/cockroachdb/pebble/1750#-N4Z-xvY-j4InlgPPftK:-N4cHDwPDtk1JW3TL0Hq:b-bw8e14) ([raw file](https://github.com/cockroachdb/pebble/blob/c2830ade5b395bc3422277c9eb1d4bbc3585f1b3/table_stats.go#L313)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

This would be range key deletions too, right? Not just range tombstones?
</blockquote></details>

Updated.
___
*[`testdata/table_stats` line 264 at r3](https://reviewable.io/reviews/cockroachdb/pebble/1750#-N4ZBCSwC9IFMORNgmht:-N4cHk3h7d4vKo2uEyOJ:b6xnxyc) ([raw file](https://github.com/cockroachdb/pebble/blob/c2830ade5b395bc3422277c9eb1d4bbc3585f1b3/testdata/table_stats#L264)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

This should be the case now, right?
</blockquote></details>

Yeah - though I have a bunch of other test cases like this (that use the ingestion hack) that I'd like to clean up. I might address these in a follow-up, to avoid the extra noise.
___
*[`testdata/table_stats_deletion_iter` line 57 at r3](https://reviewable.io/reviews/cockroachdb/pebble/1750#-N4ZBOYy7g6VoQjaJnh7:-N4cI7fDBUxBXrK-lKG-:b-jpkv8k) ([raw file](https://github.com/cockroachdb/pebble/blob/c2830ade5b395bc3422277c9eb1d4bbc3585f1b3/testdata/table_stats_deletion_iter#L57)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

Maybe we should have a case where the RANGEKEYDEL has a lower sequence number than the others to confirm this is the Filter working and not the deletion shadowing the others?
</blockquote></details>

Good call. Added.

**Review by jbowens — APPROVED** [review]
:lgtm: Looks great!

> Reviewed 1 of 7 files at r10, 2 of 5 files at r11, 1 of 2 files at r14, 2 of 2 files at r16, 1 of 1 files at r17, 1 of 1 files at r18, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1750#-:-N4da96W79JbK1jl5V5L:b-2t11ox)* status: all files reviewed, 5 unresolved discussions (waiting on @itsbilal and @nicktrav)

___
*[`compaction.go` line 1849 at r18](https://reviewable.io/reviews/cockroachdb/pebble/1750#-N4dXoq72xu6URpX5yQq:-N4dXoq8Di3FdjDD7wLF:b-hoql46) ([raw file](https://github.com/cockroachdb/pebble/blob/45aeb2c02e0027ebf8dfaf670e829be8a469c1ea/compaction.go#L1849)):*
> ```Go
> }
> 
> // compactionHintFromKeys returns a deleteCompactionHintType give a slice of
> ```

nit: s/give/given/
___
*[`table_stats.go` line 477 at r18](https://reviewable.io/reviews/cockroachdb/pebble/1750#-N4dZ_Ci2YvzxpAAYkCo:-N4dZ_Ci2YvzxpAAYkCp:bsux98k) ([raw file](https://github.com/cockroachdb/pebble/blob/45aeb2c02e0027ebf8dfaf670e829be8a469c1ea/table_stats.go#L477)):*
> ```Go
> 	// the RangeDeletionsBytesEstimate statistic and can't populate table stats
> 	// from just the properties. The table stats collector goroutine will populate
> 	// the stats.
> ```

nit: mind adjusting this comment to explicitly call out the need for the table stats collector to trigger delete-only compactions? It seems like there's two separate reasons we can't compute stats here:

1. If a table has range deletions we can't calculate `RangeDeletionsBytesEstimate` without scanning the LSM, and we need the table stats collector to do that.
2. Additionally, if a table has range deletions or range key deletions, we need the table stats collector to run on the file so that it can look for opportunities for delete-only compactions.

**Review by nicktrav — COMMENTED** [review]
Squashed all the fixup commits. I'll address the test cases in a follow-up.

TFTRs!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1750#-:-N4iQv_LBvYJk1fcJcFK:b-ub4cj6)* status: 4 of 9 files reviewed, 3 unresolved discussions (waiting on @itsbilal and @jbowens)

___
*[`compaction.go` line 1849 at r18](https://reviewable.io/reviews/cockroachdb/pebble/1750#-N4dXoq72xu6URpX5yQq:-N4iMOFH7cW-L5DWXoV_:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/45aeb2c02e0027ebf8dfaf670e829be8a469c1ea/compaction.go#L1849)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

nit: s/give/given/
</blockquote></details>

Done.
___
*[`table_stats.go` line 477 at r18](https://reviewable.io/reviews/cockroachdb/pebble/1750#-N4dZ_Ci2YvzxpAAYkCo:-N4iOpHHB6eBgzIeIyH4:b-896fja) ([raw file](https://github.com/cockroachdb/pebble/blob/45aeb2c02e0027ebf8dfaf670e829be8a469c1ea/table_stats.go#L477)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

nit: mind adjusting this comment to explicitly call out the need for the table stats collector to trigger delete-only compactions? It seems like there's two separate reasons we can't compute stats here:

1. If a table has range deletions we can't calculate `RangeDeletionsBytesEstimate` without scanning the LSM, and we need the table stats collector to do that.
2. Additionally, if a table has range deletions or range key deletions, we need the table stats collector to run on the file so that it can look for opportunities for delete-only compactions.
</blockquote></details>

Done!

**Pull request merged**
